sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -29,6 +64,476 @@ def regression_data():
+        # we get a solution, i.e. a (non-unique) minimum of the objective function ...
+        assert_allclose(model.predict(X), y, rtol=1e-6)
+        if fit_intercept:
+            # But it is not the minimum norm solution. Otherwise the norms would be
+            # equal.
+            norm_solution = np.linalg.norm(np.r_[intercept, coef])
+            norm_model = np.linalg.norm(np.r_[model.intercept_, model.coef_])
+            assert norm_model > (1 + 1e-12) * norm_solution
+
+            # Note: Even adding a tiny penalty does not give the minimal norm solution.
+            # XXX:
+            # model_pen = clone(model).set_params(**params).set_params(alpha=1e-10)
+            # model_pen.fit(X, y)
+            # assert_allclose(model_pen.predict(X), y, rtol=1e-6)  # This is true.
+            # norm_model_pen = np.linalg.norm(
+            #     np.r_[model_pen.intercept_, model_pen.coef_]
+            # )
+            # All the following assertions fail.
+            # assert norm_model_pen == pytest.approx(norm_solution)
+            # assert model_pen.intercept_ == pytest.approx(intercept)
+            # assert_allclose(model_pen.coef_, coef, rtol=rtol)
+        else:
+            assert model.intercept_ == pytest.approx(intercept)
+            assert_allclose(model.coef_, coef, rtol=rtol)
+
+
+@pytest.mark.parametrize("solver", SOLVERS)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+def test_glm_regression_unpenalized_hstacked_X(solver, fit_intercept, glm_dataset):
+    """Test that unpenalized GLM converges for all solvers to correct solution.
+
+    We work with a simple constructed data set with known solution.
+    GLM fit on [X] is the same as fit on [X, X]/2.
+    For long X, [X, X] is a singular matrix and we check against the minimum norm
+    solution:
+        min ||w||_2 subject to w = argmin deviance(X, y, w)
+    """
+    model, X, y, coef, _, _ = glm_dataset
+    n_samples, n_features = X.shape
+    alpha = 0  # unpenalized
+    params = dict(
+        alpha=alpha,
+        fit_intercept=fit_intercept,
+        # solver=solver,  # only lbfgs available
+        tol=1e-12,
+        max_iter=1000,
+    )
+
+    model = clone(model).set_params(**params)
+    if fit_intercept:
+        intercept = coef[-1]
+        coef = coef[:-1]
+        if n_samples > n_features:
+            X = X[:, :-1]  # remove intercept
+            X = 0.5 * np.concatenate((X, X), axis=1)
+        else:
+            # To know the minimum norm solution, we keep one intercept column and do
+            # not divide by 2. Later on, we must take special care.
+            X = np.c_[X[:, :-1], X[:, :-1], X[:, -1]]
+    else:
+        intercept = 0
+        X = 0.5 * np.concatenate((X, X), axis=1)
+    assert np.linalg.matrix_rank(X) <= min(n_samples, n_features)
+
+    if fit_intercept and n_samples < n_features:
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
```

Comment:
We should not set `filterwarnings` outside of a `warnings.catch_warnings` context manager to avoid having this test introducing a global side effect.

Also, is there a valid reason to expect this warning to show up in this case? If not we should at least add an inline comment with `XXX: investigate why LBFGS fails to properly convege in this case`.

Reply:
What do you propose instead of `warnings.filterwarnings`? Using a pytest decorator is not that granular. It really only happens in this case (`if` condition).

Reply:
Here is a suggestion: https://github.com/scikit-learn/scikit-learn/pull/23619/files#r900054999